src/main/java/ch/njol/skript/lang/function/FunctionRegistry.java

```diff
@@ -0,0 +1,588 @@
+package ch.njol.skript.lang.function;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.SkriptAPIException;
+import com.google.common.base.Preconditions;
+import org.jetbrains.annotations.ApiStatus.Internal;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.skriptlang.skript.lang.converter.Converters;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * A registry for functions.
+ *
+ * @author Efnilite
+ */
+@Internal // for now
+final class FunctionRegistry {
+
+	private FunctionRegistry() {
+		throw new UnsupportedOperationException("Cannot instantiate utility class");
+	}
+
+	/**
+	 * The pattern for a valid function name.
+	 * Functions must start with a letter and can only contain letters, numbers, and underscores.
+	 */
+	final static String FUNCTION_NAME_PATTERN = "\\p{IsAlphabetic}[\\p{IsAlphabetic}\\d_]*";
+
+	/**
+	 * The namespace for functions registered using Java.
+	 */
+	private static final Namespace GLOBAL_NAMESPACE = new Namespace(Scope.GLOBAL, null);
+
+	/**
+	 * Map for all function names to their identifiers, allowing for quicker lookup.
+	 */
+	private static final Map<Namespace, Map<String, Set<FunctionIdentifier>>> identifiers = new HashMap<>();
+
+	/**
+	 * Map for all identifier to function combinations, belonging to a namespace.
+	 */
+	private static final Map<Namespace, Map<FunctionIdentifier, Function<?>>> functions = new HashMap<>();
+
+	/**
+	 * Map for all identifier to signature combinations, belonging to a namespace.
+	 */
+	private static final Map<Namespace, Map<FunctionIdentifier, Signature<?>>> signatures = new HashMap<>();
```

Comment:
three different maps with the same keys implies that there should probably be a class that stores the values for a specific Namespace.